File: backend/controllers/invoiceController.js
const { ALLOWED_VAT_RATES } = require('../config/constants');

// Mock database for invoices
let mockInvoicesDb = [];
let nextInvoiceId = 1;

// Helper function to calculate invoice totals
const calculateTotals = (items, taxRate = 0) => {
    const subtotal = items.reduce((sum, item) => sum + (item.quantity * item.unit_price), 0);
    const taxAmount = subtotal * taxRate;
    const totalAmount = subtotal + taxAmount;
    return { subtotal, taxAmount, totalAmount };
};

// Helper function to generate invoice number (mock implementation)
const generateInvoiceNumber = (userId) => {
    const userInvoices = mockInvoicesDb.filter(inv => inv.user_id === userId);
    const lastInvoice = userInvoices.sort((a, b) => {
        const numA = parseInt(a.invoice_number.split('-').pop());
        const numB = parseInt(b.invoice_number.split('-').pop());
        return numA - numB;
    }).pop();

    let nextNum = 1;
    if (lastInvoice) {
        nextNum = parseInt(lastInvoice.invoice_number.split('-').pop()) + 1;
    }
    return `INV-${String(nextNum).padStart(3, '0')}`;
};


// --- Controller Functions ---

// POST /api/invoices
exports.createInvoice = (req, res) => {
    try {
        const { client_name, client_address, client_email, invoice_date, due_date, items, currency, notes, status = 'draft' } = req.body;
        let { tax_rate } = req.body; // tax_rate is handled specially for validation
        const userId = req.user.userId; // Extracted from JWT by auth middleware

        // Validation
        if (!client_name || !invoice_date || !due_date || !items || !Array.isArray(items) || items.length === 0) {
            return res.status(400).json({ message: 'Missing required fields: client_name, invoice_date, due_date, items.' });
        }
        for (const item of items) {
            if (typeof item.description !== 'string' || typeof item.quantity !== 'number' || typeof item.unit_price !== 'number') {
                return res.status(400).json({ message: 'Each item must have description (string), quantity (number), and unit_price (number).' });
            }
            item.total_price_per_item = item.quantity * item.unit_price; // Ensure total_price_per_item is set
        }

        // Validate tax_rate
        if (tax_rate === undefined || tax_rate === null) {
            tax_rate = 0; // Default to 0 if not provided
        } else {
            tax_rate = parseFloat(tax_rate); // Ensure it's a number
            if (!ALLOWED_VAT_RATES.includes(tax_rate)) {
                return res.status(400).json({
                    message: `Invalid tax_rate. Must be one of: ${ALLOWED_VAT_RATES.join(', ')}.`,
                    allowed_rates: ALLOWED_VAT_RATES
                });
            }
        }

        const { subtotal, taxAmount, totalAmount } = calculateTotals(items, tax_rate);
        const invoiceNumber = generateInvoiceNumber(userId);

        const newInvoice = {
            id: nextInvoiceId++,
            user_id: userId,
            client_name,
            client_address: client_address || null,
            client_email: client_email || null,
            invoice_number: invoiceNumber,
            invoice_date,
            due_date,
            items,
            subtotal,
            tax_rate,
            tax_amount: taxAmount,
            total_amount: totalAmount,
            currency: currency || 'USD',
            status,
            notes: notes || null,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
        };

        mockInvoicesDb.push(newInvoice);
        res.status(201).json(newInvoice);
    } catch (error) {
        console.error('Create invoice error:', error);
        res.status(500).json({ message: 'Internal server error while creating invoice.' });
    }
};

// GET /api/invoices
exports.getInvoices = (req, res) => {
    try {
        const userId = req.user.userId;
        let userInvoices = mockInvoicesDb.filter(inv => inv.user_id === userId);

        // Basic Pagination
        const page = parseInt(req.query.page, 10) || 1;
        const limit = parseInt(req.query.limit, 10) || 10;
        const startIndex = (page - 1) * limit;
        const endIndex = page * limit;

        const paginatedInvoices = userInvoices.slice(startIndex, endIndex);

        res.status(200).json({
            page,
            limit,
            totalPages: Math.ceil(userInvoices.length / limit),
            totalItems: userInvoices.length,
            invoices: paginatedInvoices
        });
    } catch (error) {
        console.error('Get invoices error:', error);
        res.status(500).json({ message: 'Internal server error while fetching invoices.' });
    }
};

// GET /api/invoices/:id
exports.getInvoiceById = (req, res) => {
    try {
        const userId = req.user.userId;
        const invoiceId = parseInt(req.params.id, 10);
        const invoice = mockInvoicesDb.find(inv => inv.id === invoiceId);

        if (!invoice) {
            return res.status(404).json({ message: 'Invoice not found.' });
        }
        if (invoice.user_id !== userId) {
            // Although the list endpoint already filters by user, this is a direct access check
            return res.status(403).json({ message: 'Forbidden: You do not have access to this invoice.' });
        }

        res.status(200).json(invoice);
    } catch (error) {
        console.error('Get invoice by ID error:', error);
        res.status(500).json({ message: 'Internal server error.' });
    }
};

// PUT /api/invoices/:id
exports.updateInvoice = (req, res) => {
    try {
        const userId = req.user.userId;
        const invoiceId = parseInt(req.params.id, 10);
        const updates = req.body;
        let { tax_rate: newTaxRate } = req.body; // Handle new tax_rate separately

        let invoice = mockInvoicesDb.find(inv => inv.id === invoiceId);

        if (!invoice) {
            return res.status(404).json({ message: 'Invoice not found.' });
        }
        if (invoice.user_id !== userId) {
            return res.status(403).json({ message: 'Forbidden: You cannot update this invoice.' });
        }

        // Fields that cannot be changed
        delete updates.id;
        delete updates.user_id;
        delete updates.invoice_number; // Invoice number should generally not be changed post-creation
        delete updates.created_at;

        // Update allowed fields
        Object.keys(updates).forEach(key => {
            if (key !== 'items' && key !== 'tax_rate' && invoice.hasOwnProperty(key)) {
                invoice[key] = updates[key];
            }
        });

        // Handle updates to items or tax_rate which require recalculation
        let needsRecalculation = false;
        if (updates.items && Array.isArray(updates.items) && updates.items.length > 0) {
            // Basic validation for incoming items
            for (const item of updates.items) {
                if (typeof item.description !== 'string' || typeof item.quantity !== 'number' || typeof item.unit_price !== 'number') {
                    return res.status(400).json({ message: 'Each updated item must have description (string), quantity (number), and unit_price (number).' });
                }
                item.total_price_per_item = item.quantity * item.unit_price;
            }
            invoice.items = updates.items;
            needsRecalculation = true;
        }

        // Validate and update tax_rate if provided
        if (newTaxRate !== undefined && newTaxRate !== null) {
            newTaxRate = parseFloat(newTaxRate);
            if (!ALLOWED_VAT_RATES.includes(newTaxRate)) {
                return res.status(400).json({
                    message: `Invalid tax_rate. Must be one of: ${ALLOWED_VAT_RATES.join(', ')}.`,
                    allowed_rates: ALLOWED_VAT_RATES
                });
            }
            if (invoice.tax_rate !== newTaxRate) {
                 invoice.tax_rate = newTaxRate;
                 needsRecalculation = true;
            }
        } else if (updates.hasOwnProperty('tax_rate') && (updates.tax_rate === null || updates.tax_rate === undefined)) {
            // If explicitly set to null/undefined, treat as 0 or reset, and recalculate
            if (invoice.tax_rate !== 0) {
                invoice.tax_rate = 0;
                needsRecalculation = true;
            }
        }


        if (needsRecalculation) {
            const { subtotal, taxAmount, totalAmount } = calculateTotals(invoice.items, invoice.tax_rate);
            invoice.subtotal = subtotal;
            invoice.tax_amount = taxAmount;
            invoice.total_amount = totalAmount;
        }

        invoice.updated_at = new Date().toISOString();

        // Replace in DB
        mockInvoicesDb = mockInvoicesDb.map(inv => (inv.id === invoiceId ? invoice : inv));

        res.status(200).json(invoice);
    } catch (error) {
        console.error('Update invoice error:', error);
        res.status(500).json({ message: 'Internal server error while updating invoice.' });
    }
};

// DELETE /api/invoices/:id
exports.deleteInvoice = (req, res) => {
    try {
        const userId = req.user.userId;
        const invoiceId = parseInt(req.params.id, 10);

        const invoiceIndex = mockInvoicesDb.findIndex(inv => inv.id === invoiceId);

        if (invoiceIndex === -1) {
            return res.status(404).json({ message: 'Invoice not found.' });
        }
        if (mockInvoicesDb[invoiceIndex].user_id !== userId) {
            return res.status(403).json({ message: 'Forbidden: You cannot delete this invoice.' });
        }

        mockInvoicesDb.splice(invoiceIndex, 1);
        res.status(204).send();
    } catch (error) {
        console.error('Delete invoice error:', error);
        res.status(500).json({ message: 'Internal server error while deleting invoice.' });
    }
};

// Export for use in other modules (like paymentController)
// This is a simplified approach for a mock environment.
// In a real app, you'd use a proper database and service layer.
module.exports.mockInvoicesDb = mockInvoicesDb;
module.exports.updateInvoiceStatus = (invoiceId, newStatus) => {
    const invoice = mockInvoicesDb.find(inv => inv.id === invoiceId);
    if (invoice) {
        invoice.status = newStatus;
        invoice.updated_at = new Date().toISOString();
        // In a real scenario, you might want to re-save/update the DB
        // For mockInvoicesDb, the change is in-memory.
        return true;
    }
    return false;
};

File: backend/controllers/invoice_controller.py
# Conceptual Implementation for Invoice CRUD API Endpoints
# Assumes integration with a web framework (e.g., Flask, Django, FastAPI)
# and use of models from backend.models.invoice (e.g., ALLOWED_VAT_RATES).

# Important Considerations:
# 1. Input Validation: All incoming data must be validated (e.g., data types, required fields, formats).
#    - `tax_rate` must be one of the `ALLOWED_VAT_RATES` from `invoice.py`.
# 2. Authorization:
#    - Users should only be able to access/modify their own invoices unless they have specific roles (e.g., accountant, admin).
#    - RBAC (Role-Based Access Control) checks from `rbac_service.py` should be integrated.
# 3. Error Handling: Consistent error responses should be provided.
# 4. Database Interaction: These conceptual functions would interact with a database layer.
# 5. Invoice Number Generation: Logic for generating unique invoice numbers is needed.
# 6. Client Management: `client_id` would ideally link to a separate `Clients` table/model.
# 7. "TVA non applicable" mention:
#    - If `tax_rate` is 0.00, the system might need to generate a specific legal mention on the final invoice PDF
#      (e.g., "TVA non applicable, art. 293B du CGI" for France). This depends on the user's country
#      and tax status. This logic is typically part of the PDF generation/presentation layer.

# --- API Endpoints ---

# 1. Create a New Invoice
# Endpoint: POST /api/invoices
# Description: Creates a new invoice.
# Request Payload (JSON):
# {
#   "client_id": "client_uuid_or_id", # Or "client_name", "client_address"
#   "client_name": "Example Client Inc.",
#   "client_address": "456 Business Rd, Suite 789, Cityville",
#   "issue_date": "2024-07-31", # ISO format YYYY-MM-DD
#   "due_date": "2024-08-30",   # ISO format YYYY-MM-DD
#   "items": [ # Item prices are HT (before tax)
#     {"description": "Product A (HT)", "quantity": 2, "unit_price": 50.00}, # total_price HT = 100.00
#     {"description": "Service B (HT)", "quantity": 1, "unit_price": 200.00} # total_price HT = 200.00
#   ],
#   "currency": "USD", # "XOF", "EUR"
#   "tax_rate": 0.08,  # 8%. Must be one of ALLOWED_VAT_RATES.
#   "notes": "Payment due within 30 days.",
#   "status": "draft" # Optional, defaults to 'draft'
# }
# Conceptual Logic during creation:
# - Sum of `item.total_price` gives `subtotal` (e.g., 100.00 + 200.00 = 300.00 HT).
# - `tax_amount` = `subtotal` * `tax_rate` (e.g., 300.00 * 0.08 = 24.00).
# - `total_amount` = `subtotal` + `tax_amount` (e.g., 300.00 + 24.00 = 324.00 TTC).
# Response Payload (Success - 201 Created):
# {
#   "invoice_id": "new_invoice_uuid_or_id",
#   "user_id": "creator_user_id",
#   "invoice_number": "INV-2024-001", # Generated invoice number
#   "subtotal": 300.00,
#   "tax_rate": 0.08,
#   "tax_amount": 24.00,
#   "total_amount": 324.00,
#   "currency": "USD",
#   "status": "draft",
#   "message": "Invoice created successfully."
#   # ... other relevant fields from the created invoice
# }
# Response Payload (Error - 400 Bad Request):
# { "error": "Invalid input data. Missing required fields, incorrect formats, or invalid tax_rate." }
# Response Payload (Error - 403 Forbidden):
# { "error": "You do not have permission to create invoices." }

# 2. List All Invoices (for the authenticated user)
# Endpoint: GET /api/invoices
# Description: Retrieves a list of invoices for the authenticated user.
#             Supports filtering (e.g., by status, client, date range) and pagination.
# Query Parameters (Optional):
#   - `status` (e.g., 'draft', 'sent', 'paid')
#   - `client_id`
#   - `date_from`, `date_to`
#   - `page` (for pagination)
#   - `limit` (for pagination)
# Response Payload (Success - 200 OK):
# {
#   "data": [
#     {
#       "invoice_id": "inv_id_1",
#       "invoice_number": "INV-2024-001",
#       "client_name": "Client A",
#       "issue_date": "2024-07-15",
#       "due_date": "2024-08-14",
#       "total_amount": 162.00, # Example total TTC
#       "currency": "USD",
#       "status": "sent"
#     },
#     {
#       "invoice_id": "inv_id_2",
#       "invoice_number": "INV-2024-002",
#       "client_name": "Client B",
#       "issue_date": "2024-07-20",
#       "due_date": "2024-08-19",
#       "total_amount": 300.00, # Example total TTC
#       "currency": "XOF",
#       "status": "paid"
#     }
#   ],
#   "pagination": {
#     "total_items": 50,
#     "total_pages": 5,
#     "current_page": 1,
#     "page_size": 10
#   }
# }
# Response Payload (Error - 403 Forbidden):
# { "error": "You do not have permission to view these invoices." }


# 3. Get a Specific Invoice
# Endpoint: GET /api/invoices/{invoice_id}
# Description: Retrieves details of a specific invoice.
# Response Payload (Success - 200 OK):
# {
#   "invoice_id": "inv_id_1",
#   "user_id": "creator_user_id",
#   "invoice_number": "INV-2024-001",
#   "client_name": "Client A",
#   "client_address": "123 Main St",
#   "issue_date": "2024-07-15",
#   "due_date": "2024-08-14",
#   "items": [ # Item prices are HT
#     {"description": "Item 1 (HT)", "quantity": 1, "unit_price": 100.00, "total_price": 100.00}
#   ],
#   "subtotal": 100.00, # Sum of HT item totals
#   "tax_rate": 0.10,   # Document-level tax rate applied
#   "tax_amount": 10.00, # subtotal * tax_rate
#   "total_amount": 110.00, # subtotal + tax_amount (TTC)
#   "currency": "USD",
#   "status": "sent",
#   "notes": "Thanks!",
#   "created_at": "2024-07-15T10:00:00Z",
#   "updated_at": "2024-07-15T10:00:00Z"
# }
# Response Payload (Error - 403 Forbidden):
# { "error": "You do not have permission to view this invoice." }
# Response Payload (Error - 404 Not Found):
# { "error": "Invoice not found." }

# 4. Update an Existing Invoice
# Endpoint: PUT /api/invoices/{invoice_id}
# Description: Updates an existing invoice. Only certain fields might be updatable depending on status (e.g., a 'paid' invoice might be locked).
# Request Payload (JSON): Contains fields to be updated.
# {
#   "client_name": "Updated Client A",
#   "due_date": "2024-08-20",
#   "items": [ # Item prices are HT
#     {"description": "Item 1 (updated HT)", "quantity": 1, "unit_price": 120.00}
#   ],
#   "tax_rate": 0.05, # Optional: if changing tax rate. Must be one of ALLOWED_VAT_RATES.
#   "notes": "Updated payment terms.",
#   "status": "draft" # Potentially
# }
# Conceptual Logic during update:
# - If `items` are changed, `subtotal` is recalculated from new item HT totals.
# - If `tax_rate` is changed (and valid), it's updated.
# - `tax_amount` = new `subtotal` * new or existing `tax_rate`.
# - `total_amount` = new `subtotal` + new `tax_amount`.
# Response Payload (Success - 200 OK):
# {
#   "invoice_id": "inv_id_1",
#   "message": "Invoice updated successfully.",
#   "subtotal": 120.00,
#   "tax_rate": 0.05,
#   "tax_amount": 6.00,
#   "total_amount": 126.00,
#   # ... other relevant fields from the updated invoice
# }
# Response Payload (Error - 400 Bad Request):
# { "error": "Invalid input data or invalid tax_rate." }
# Response Payload (Error - 403 Forbidden):
# { "error": "You do not have permission to update this invoice." }
# Response Payload (Error - 404 Not Found):
# { "error": "Invoice not found." }
# Response Payload (Error - 409 Conflict):
# { "error": "Cannot update invoice in its current state (e.g., if paid)." }

# 5. Delete an Invoice
# Endpoint: DELETE /api/invoices/{invoice_id}
# Description: Deletes an invoice (soft delete is often preferred over hard delete).
# Response Payload (Success - 200 OK or 204 No Content):
# { "message": "Invoice deleted successfully." }
# Response Payload (Error - 403 Forbidden):
# { "error": "You do not have permission to delete this invoice." }
# Response Payload (Error - 404 Not Found):
# { "error": "Invoice not found." }
# Response Payload (Error - 409 Conflict):
# { "error": "Cannot delete invoice in its current state (e.g., if paid or part of an audit trail)." }

# --- Placeholder Functions (to be implemented with a framework) ---

def create_invoice_handler(request_data, user_id):
    # 1. Validate input_data (using a schema validator like Pydantic or Marshmallow)
    # 2. Check permissions (e.g., rbac_service.check_permission(user.role, PERMISSIONS["CREATE_INVOICE"]))
    # 3. Generate invoice_number
    # 4. Create InvoiceLineItem objects from request_data.items
    # 5. Create Invoice object (from models.invoice)
    # 6. Save to database
    # 7. Return response
    pass

def list_invoices_handler(user_id, query_params):
    # 1. Check permissions (e.g., rbac_service.check_permission(user.role, PERMISSIONS["VIEW_OWN_INVOICES"]))
    #    Accountants/Admins might have PERMISSIONS["VIEW_ALL_INVOICES"]
    # 2. Construct database query based on user_id and query_params (filters, pagination)
    # 3. Fetch invoices from database
    # 4. Format response
    pass

def get_invoice_handler(invoice_id, user_id):
    # 1. Fetch invoice from database by invoice_id
    # 2. If not found, return 404
    # 3. Check ownership or specific view permissions (e.g., user_id matches invoice.user_id OR rbac allows viewing all)
    #    if not (invoice.user_id == user_id or rbac_service.check_permission(user.role, PERMISSIONS["VIEW_ALL_INVOICES"])):
    #        return {"error": "Forbidden"}, 403
    # 4. Format and return response
    pass

def update_invoice_handler(invoice_id, request_data, user_id):
    # 1. Fetch invoice from database
    # 2. If not found, return 404
    # 3. Check ownership and permissions
    # 4. Validate input_data
    # 5. Check if invoice status allows updates (e.g., not 'paid' or 'cancelled')
    # 6. Update invoice fields, recalculate totals if items change
    # 7. Save to database
    # 8. Format and return response
    pass

def delete_invoice_handler(invoice_id, user_id):
    # 1. Fetch invoice from database
    # 2. If not found, return 404
    # 3. Check ownership and permissions
    # 4. Check if invoice status allows deletion (e.g., not 'paid')
    # 5. Perform soft delete (e.g., set status to 'deleted' or mark is_deleted=True) or hard delete
    # 6. Return success response
    pass
